SeQuant/domain/eval/eval_result.hpp

```diff
@@ -0,0 +1,684 @@
+  [[nodiscard]] bool is() const noexcept {
+    return this->type_id() == id_for_type<std::decay_t<T>>();
+  }
+
+  template <typename T>
+  [[nodiscard]] T const& as() const {
+    assert(this->is<std::decay_t<T>>());
+    return static_cast<T const&>(*this);
+  }
+
+  ///
+  /// @note In std::array<std::any, 3> is expected to be [l,r,res] where
+  ///       the elements are the annotations for left, right and result
+  ///       respectively.
+  ///
+  [[nodiscard]] virtual ERPtr sum(EvalResult const&,
+                                  std::array<std::any, 3> const&) const = 0;
+
+  ///
+  /// @note In std::array<std::any, 3> is expected to be [l,r,res] where
+  ///       the elements are the annotations for left, right and result
+  ///       respectively.
+  ///
+  [[nodiscard]] virtual ERPtr prod(EvalResult const&,
+                                   std::array<std::any, 3> const&) const = 0;
+
+  ///
+  /// @note In std::array<std::any, 2> is expected to be [pre,post] where
+  ///       the elements are the annotations for the eval result before
+  ///       permutation and after permutation respectively.
+  ///
+  [[nodiscard]] virtual ERPtr permute(std::array<std::any, 2> const&) const = 0;
+
+  virtual void add_inplace(EvalResult const&) = 0;
+
+  ///
+  /// @note vector<array<size_t,3>> represents list of particle
+  ///       symmetry index groups. array<size_t, 3> is expected to be
+  ///         [b1, b2, len]
+  ///       where b1, and b2 are the zero-based positions of the tensor indices.
+  ///       [b1, b1+len) and [b2, b2+len) are two ranges that will be permuted
+  ///       simultaneously and at the equivalent positions.
+  ///
+  [[nodiscard]] virtual ERPtr symmetrize(
+      container::svector<std::array<size_t, 3>> const&) const = 0;
+
+  ///
+  /// @note vector<array<size_t,3>> represents list of antisymmetric index
+  ///       groups. array<size_t,3> is expected to be
+  ///         [b1,b2,len]
+  ///       where b1, and b2 are the zero-based positions of the tensor indices.
+  ///       [b1, b1+len) and [b2, b2+len) are two ranges that will be permuted
+  ///       by keeping track of the parity (even/odd)-ness of the total
+  ///       permutation.
+  ///
+  [[nodiscard]] virtual ERPtr antisymmetrize(
+      container::svector<std::array<size_t, 3>> const&) const = 0;
+
+  [[nodiscard]] bool has_value() const noexcept;
+
+  template <typename T>
+  [[nodiscard]] T& get() {
+    assert(has_value());
+    return *std::any_cast<T>(&value_);
+  }
+
+  template <typename T>
+  [[nodiscard]] T const& get() const {
+    return const_cast<EvalResult&>(*this).get<T>();
+  }
+
+ protected:
+  template <typename T>
+  [[nodiscard]] T& get_ref() {
+    assert(has_value());
+    return *std::any_cast<T>(&value_);
+  }
+
+  template <typename T,
+            typename = std::enable_if_t<!std::is_convertible_v<T, EvalResult>>>
+  explicit EvalResult(T&& arg) noexcept
+      : value_{std::make_any<T>(std::forward<T>(arg))} {}
```

Comment:
`T` can be deduced to be a reference type here, need to `make_any<std::decay<T>>(`